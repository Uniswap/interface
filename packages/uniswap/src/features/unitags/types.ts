// Internal types

export type UnitagClaim = {
  address?: string
  username: string
  avatarUri?: string
}

export type UnitagClaimSource = 'onboarding' | 'home' | 'settings'

export type UnitagClaimContext = {
  source: UnitagClaimSource
  hasENSAddress: boolean
}

// API types

export type UnitagUsernameRequest = {
  username: string
}

export type UnitagUsernameResponse = {
  available: boolean
  requiresEnsMatch: boolean
  metadata?: ProfileMetadata
  username?: string
  address?: {
    address: Address
  }
}

export type UnitagAddressRequest = {
  address: string
}

export type UnitagAddressResponse = {
  username?: string
  address?: Address
  metadata?: ProfileMetadata
}

export type UnitagAddressesRequest = {
  addresses: Address[]
}

export type UnitagAddressesResponse = {
  usernames: {
    [address: Address]: UnitagAddressResponse
  }
}

export type UnitagResponse = {
  success: boolean
  errorCode?: UnitagErrorCodes
}

export type UnitagClaimEligibilityResponse = {
  canClaim: boolean
  errorCode?: UnitagErrorCodes
}

export type ProfileMetadata = {
  avatar?: string
  description?: string
  twitter?: string
}

export type UnitagClaimUsernameRequestBody = {
  username: string
  deviceId: string
  metadata?: ProfileMetadata
}

export type UnitagClaimEligibilityRequest = {
  address?: Address
  deviceId: string
  isUsernameChange?: boolean
}

export type UnitagUpdateMetadataRequestBody = {
  metadata: ProfileMetadata
  clearAvatar?: boolean
}

export type UnitagUpdateMetadataResponse = {
  success: boolean
  metadata?: ProfileMetadata
}

export type UnitagGetAvatarUploadUrlResponse = {
  success: boolean
  avatarUrl?: string
  preSignedUrl?: string
  s3UploadFields?: Record<string, string>
}

export type UnitagDeleteUsernameRequestBody = {
  username: string
}

export type UnitagAvatarUploadCredentials = {
  preSignedUrl?: string
  s3UploadFields?: Record<string, string>
}

export type UnitagChangeUsernameRequestBody = {
  username: string
  deviceId: string
}

// Copied enum from unitags backend code -- needs to be up-to-date
export enum UnitagErrorCodes {
  UnitagNotAvailable = 'unitags-1',
  RequiresENSMatch = 'unitags-2',
  IPLimitReached = 'unitags-3',
  AddressLimitReached = 'unitags-4',
  DeviceLimitReached = 'unitags-5',
  DeviceActiveLimitReached = 'unitags-6',
  AddressActiveLimitReached = 'unitags-7',
  NoUnitagForAddress = 'unitags-8',
  UnitagNotActive = 'unitags-9',
}
