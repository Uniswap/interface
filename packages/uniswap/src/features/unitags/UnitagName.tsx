import { Flex, FlexProps, Text } from 'ui/src'
import { Unitag } from 'ui/src/components/icons/Unitag'
import { fonts, spacing } from 'ui/src/theme'

export function UnitagName({
  name,
  fontSize,
  opacity = 1,
  animateIcon = false,
  displayIconInline = false,
}: {
  name?: string
  fontSize: number
  opacity?: number
  animateIcon?: boolean
  displayIconInline?: boolean
}): JSX.Element {
  const iconContainerProps: FlexProps = displayIconInline
    ? {}
    : {
        position: 'absolute',
        right: -spacing.spacing24,
        top: -spacing.spacing4,
      }

  return (
    <Flex
      row
      alignSelf="center"
      animation="lazy"
      enterStyle={{ opacity: 0 }}
      exitStyle={{ opacity: 0 }}
      opacity={opacity}
      alignItems="center"
      gap="$spacing4"
    >
      <Text
        color="$neutral1"
        fontFamily="$heading"
        fontSize={fontSize}
        fontWeight={fonts.heading2.fontWeight}
        lineHeight={fonts.heading2.lineHeight}
      >
        {name}
      </Text>
      <Flex
        {...iconContainerProps}
        row
        animation="lazy"
        enterStyle={animateIcon ? { opacity: 0, scale: 0.8, x: 20 } : undefined}
        exitStyle={animateIcon ? { opacity: 0, scale: 0.8, x: -20 } : undefined}
      >
        <Unitag size="$icon.24" />
      </Flex>
    </Flex>
  )
}
