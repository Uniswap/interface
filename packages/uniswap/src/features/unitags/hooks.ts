import { useUnitagsAddressQuery } from 'uniswap/src/data/apiClients/unitagsApi/useUnitagsAddressQuery'
import { useUnitagsUsernameQuery } from 'uniswap/src/data/apiClients/unitagsApi/useUnitagsUsernameQuery'
import { UnitagAddressResponse, UnitagUsernameResponse } from 'uniswap/src/features/unitags/types'

// TODO(WALL-4256): delete these helpers and just use `useUnitagsAddressQuery` and `useUnitagsUsernameQuery` where needed.

export const useUnitagByAddress = (
  address?: Address,
): { unitag?: UnitagAddressResponse; loading: boolean; pending: boolean; fetching: boolean } => {
  const { data, isLoading, isPending, isFetching } = useUnitagsAddressQuery({
    params: address ? { address } : undefined,
  })
  return { unitag: data, loading: isLoading, pending: isPending, fetching: isFetching }
}

export const useUnitagByName = (username?: string): { unitag?: UnitagUsernameResponse; loading: boolean } => {
  const { data, isLoading } = useUnitagsUsernameQuery({ params: username ? { username } : undefined })
  return { unitag: data, loading: isLoading }
}
