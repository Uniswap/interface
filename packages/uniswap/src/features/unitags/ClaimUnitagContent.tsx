import { EventConsumer, EventMapBase } from '@react-navigation/core'
import { ADDRESS_ZERO } from '@uniswap/v3-sdk'
import { useCallback, useEffect, useRef, useState } from 'react'
import { useTranslation } from 'react-i18next'
import type { LayoutChangeEvent } from 'react-native'
import { useAnimatedStyle, useSharedValue, withDelay, withTiming } from 'react-native-reanimated'
import { AnimatePresence, Button, Flex, FlexProps, Input, Text, TouchableArea } from 'ui/src'
import { CheckmarkCircle } from 'ui/src/components/icons/CheckmarkCircle'
import { InfoCircleFilled } from 'ui/src/components/icons/InfoCircleFilled'
import { AnimatedFlex } from 'ui/src/components/layout/AnimatedFlex'
import { useDynamicFontSizing } from 'ui/src/hooks/useDynamicFontSizing'
import { fonts, imageSizes, spacing } from 'ui/src/theme'
import { TextInput } from 'uniswap/src/components/input/TextInput'
import { UnitagEventName } from 'uniswap/src/features/telemetry/constants'
import { sendAnalyticsEvent } from 'uniswap/src/features/telemetry/send'
import { UnitagInfoModal } from 'uniswap/src/features/unitags/UnitagInfoModal'
import { UnitagName } from 'uniswap/src/features/unitags/UnitagName'
import { UNITAG_SUFFIX } from 'uniswap/src/features/unitags/constants'
import { getUnitagFormatError, useCanClaimUnitagName } from 'uniswap/src/features/unitags/hooks/useCanClaimUnitagName'
import { getYourNameString } from 'uniswap/src/features/unitags/utils'
import { TestID } from 'uniswap/src/test/fixtures/testIDs'
import {
  OnboardingScreens,
  SharedUnitagScreenParams,
  UnitagEntryPoint,
  UnitagScreens,
} from 'uniswap/src/types/screens/mobile'
import { shortenAddress } from 'utilities/src/addresses'
import { dismissNativeKeyboard } from 'utilities/src/device/keyboard/dismissNativeKeyboard'
import { logger } from 'utilities/src/logger/logger'
import { isMobileApp, isWeb } from 'utilities/src/platform'
import { useEvent } from 'utilities/src/react/hooks'
import { ONE_SECOND_MS } from 'utilities/src/time/time'
import { useDebounce } from 'utilities/src/time/timing'

const VERIFICATION_DEBOUNCE_MS = 700
const MAX_UNITAG_CHAR_LENGTH = 20

const MAX_INPUT_FONT_SIZE = 36
const MIN_INPUT_FONT_SIZE = 22
const MAX_CHAR_PIXEL_WIDTH = 20
const SLIDE_IN_AMOUNT = isWeb ? 0 : 40

// Used in dynamic font size width calculation to ignore `.` characters
const UNITAG_SUFFIX_CHARS_ONLY = UNITAG_SUFFIX.replaceAll('.', '')

// Accounts for height of image, gap between image and name, and spacing from top of titles
const UNITAG_NAME_ANIMATE_DISTANCE_Y = imageSizes.image100 + spacing.spacing48 + spacing.spacing24

export type ClaimUnitagContentProps = {
  unitagAddress?: string
  entryPoint: UnitagEntryPoint
  animateY?: boolean
  navigationEventConsumer?: EventConsumer<EventMapBase>
  onNavigateContinue?: (params: SharedUnitagScreenParams[UnitagScreens.ChooseProfilePicture]) => void
  onComplete?: (unitag: string) => void
}

export function ClaimUnitagContent({
  unitagAddress,
  entryPoint,
  animateY = true,
  navigationEventConsumer,
  onNavigateContinue,
  onComplete,
}: ClaimUnitagContentProps): JSX.Element {
  const { t } = useTranslation()
  const textInputRef = useRef<Input>(null)

  const inputPlaceholder = getYourNameString(t('unitags.claim.username.default'))

  const [showInfoModal, setShowInfoModal] = useState(false)

  const [showTextInputView, setShowTextInputView] = useState(true)
  const [unitagInputValue, setUnitagInputValue] = useState<string | undefined>(undefined)
  const [isUnitagAvailable, setIsUnitagAvailable] = useState(false)
  const [unitagAvailableError, setUnitagAvailableError] = useState<string>()
  const [showVerificationLoading, setShowVerificationLoading] = useState(false)

  const [unitagNameinputMinWidth, setUnitagNameInputMinWidth] = useState<number | undefined>(undefined)
  const [addressError, setAddressError] = useState<string>()

  const addressViewOpacity = useSharedValue(1)
  const unitagInputContainerTranslateY = useSharedValue(0)
  const addressViewAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: addressViewOpacity.value,
    }
  }, [addressViewOpacity])

  const debouncedInputValue = useDebounce(unitagInputValue, VERIFICATION_DEBOUNCE_MS)
  const { error: canClaimUnitagNameError, loading: isCheckingUnitag } = useCanClaimUnitagName(
    debouncedInputValue || undefined, // set to undefined if the input is empty to clear the error
  )

  const { onLayout, fontSize, onSetFontSize } = useDynamicFontSizing(
    MAX_CHAR_PIXEL_WIDTH,
    MAX_INPUT_FONT_SIZE,
    MIN_INPUT_FONT_SIZE,
  )

  const focusUnitagTextInput = useCallback((): void | null => {
    textInputRef.current?.focus()
  }, [textInputRef])

  const handleHideInfoModal = useCallback(() => {
    setShowInfoModal(false)
  }, [])

  useEffect(() => {
    return navigationEventConsumer?.addListener('transitionEnd', focusUnitagTextInput)
  }, [navigationEventConsumer, focusUnitagTextInput])

  useEffect(() => {
    const unsubscribe = navigationEventConsumer?.addListener('focus', () => {
      // When returning back to this screen, handle animating the Unitag logo out and text input in
      if (showTextInputView) {
        return
      }

      unitagInputContainerTranslateY.value = withTiming(
        unitagInputContainerTranslateY.value - UNITAG_NAME_ANIMATE_DISTANCE_Y,
        {
          duration: ONE_SECOND_MS / 2,
        },
      )
      setTimeout(() => {
        setShowTextInputView(true)
        addressViewOpacity.value = withTiming(1, { duration: ONE_SECOND_MS / 2 })
        focusUnitagTextInput()
      }, ONE_SECOND_MS)
    })

    return unsubscribe
  }, [
    navigationEventConsumer,
    showTextInputView,
    setShowTextInputView,
    addressViewOpacity,
    unitagInputContainerTranslateY,
    focusUnitagTextInput,
  ])

  const onChangeTextInput = useCallback(
    (text: string): void => {
      setIsUnitagAvailable(false)
      setShowVerificationLoading(false)
      setUnitagAvailableError(undefined)

      if (text.length > MAX_UNITAG_CHAR_LENGTH) {
        setUnitagAvailableError(getUnitagFormatError(text, t))
        return
      }

      if (text.length === 0) {
        onSetFontSize(inputPlaceholder + UNITAG_SUFFIX_CHARS_ONLY)
      } else {
        onSetFontSize(text + UNITAG_SUFFIX_CHARS_ONLY)
      }

      setUnitagInputValue(text?.trim())
    },
    [inputPlaceholder, onSetFontSize, t],
  )

  const onPressAddressTooltip = useCallback((): void => {
    dismissNativeKeyboard()
    setShowInfoModal(true)
  }, [])

  const navigateWithAnimation = useCallback(
    (unitag: string) => {
      // Log claim display and action taken
      sendAnalyticsEvent(UnitagEventName.UnitagClaimAvailabilityDisplayed, {
        result: 'available',
      })
      sendAnalyticsEvent(UnitagEventName.UnitagOnboardingActionTaken, { action: 'select' })

      // Animate the Unitag logo in and text input out
      setShowTextInputView(false)

      const initialDelay = ONE_SECOND_MS
      const translateYDuration = ONE_SECOND_MS / 2

      addressViewOpacity.value = withTiming(0, { duration: ONE_SECOND_MS / 2 })
      // Intentionally delay 1s to allow enter/exit animations to finish
      unitagInputContainerTranslateY.value = withDelay(
        initialDelay,
        withTiming(unitagInputContainerTranslateY.value + UNITAG_NAME_ANIMATE_DISTANCE_Y, {
          duration: translateYDuration,
        }),
      )
      // Navigate to ChooseProfilePicture screen after initial delay + translation to allow animations to finish
      setTimeout(() => {
        onComplete?.(unitag)
        if (unitagAddress && onNavigateContinue) {
          onNavigateContinue({ unitag, entryPoint, address: unitagAddress, unitagFontSize: fontSize })
        }
      }, initialDelay + translateYDuration)
    },
    [
      onComplete,
      onNavigateContinue,
      addressViewOpacity,
      entryPoint,
      unitagAddress,
      unitagInputContainerTranslateY,
      fontSize,
    ],
  )

  useEffect(() => {
    if (!!debouncedInputValue && !isCheckingUnitag) {
      // If unitagError or addressError is defined, it's rendered in UI
      if (entryPoint === OnboardingScreens.Landing && !unitagAddress) {
        const err = new Error('unitagAddress should always be defined')
        logger.error(err, {
          tags: { file: 'ClaimUnitagScreen', function: 'navigateWithAnimation' },
        })
        setAddressError(t('unitags.claim.error.default'))
        setIsUnitagAvailable(false)
        return
      } else if (!canClaimUnitagNameError) {
        setIsUnitagAvailable(true)
      } else {
        sendAnalyticsEvent(UnitagEventName.UnitagClaimAvailabilityDisplayed, {
          result: 'unavailable',
        })
        setIsUnitagAvailable(false)
        setUnitagAvailableError(canClaimUnitagNameError)
      }
    }
  }, [canClaimUnitagNameError, debouncedInputValue, isCheckingUnitag, entryPoint, unitagAddress, t])

  const shouldBlockContinue = (entryPoint === OnboardingScreens.Landing && !unitagAddress) || !unitagInputValue

  const onPressContinue = useEvent((): void => {
    if (isCheckingUnitag) {
      setShowVerificationLoading(true)
      return
    }

    if (unitagInputValue && isUnitagAvailable) {
      navigateWithAnimation(unitagInputValue)
    }
  })

  const webStyling: FlexProps = isWeb
    ? {
        backgroundColor: '$surface1',
        borderRadius: '$rounded20',
        borderWidth: 1,
        borderColor: '$surface3',
        py: '$spacing12',
        px: '$spacing20',
        mb: '$spacing20',
        width: '100%',
        justifyContent: 'space-between',
      }
    : {}

  const getInitialUnitagNameInputWidth = (event: LayoutChangeEvent): void => {
    if (unitagNameinputMinWidth) {
      return
    }

    // Fix from WALL-4822 for Android
    // Sets input minWidth to initial input width + 1 point. Initial width is not sufficient after clearing the input.
    setUnitagNameInputMinWidth(event.nativeEvent.layout.width + 1)
  }

  return (
    <>
      <Flex
        centered
        gap="$spacing12"
        mt="$spacing24"
        onLayout={(event): void => {
          onLayout(event)
          onSetFontSize(inputPlaceholder + UNITAG_SUFFIX_CHARS_ONLY)
        }}
      >
        {/* Fixed text that animates in when TextInput is animated out */}
        <AnimatedFlex
          centered
          width="100%"
          height={fonts.heading2.lineHeight}
          style={{ transform: [{ translateY: animateY ? unitagInputContainerTranslateY : 0 }] }}
        >
          {!showTextInputView && (
            <Flex position="absolute">
              <UnitagName
                animateText
                animateIcon
                textProps={{ fontSize }}
                name={unitagInputValue}
                opacity={showTextInputView ? 0 : 1}
              />
            </Flex>
          )}
          <AnimatePresence>
            {showTextInputView && (
              <Flex
                key="input-container"
                row
                animation="quick"
                enterStyle={{ opacity: 0, x: SLIDE_IN_AMOUNT }}
                exitStyle={{ opacity: 0, x: SLIDE_IN_AMOUNT }}
                gap="$none"
                {...webStyling}
              >
                <TextInput
                  ref={textInputRef}
                  autoFocus={!isMobileApp}
                  blurOnSubmit={!isWeb}
                  autoCapitalize="none"
                  autoCorrect={false}
                  borderWidth="$none"
                  borderRadius={isWeb ? 0 : undefined}
                  fontFamily="$heading"
                  fontSize={isWeb ? fonts.subheading1.fontSize : fontSize}
                  fontWeight="$book"
                  numberOfLines={1}
                  p="$none"
                  placeholder={inputPlaceholder}
                  placeholderTextColor="$neutral3"
                  returnKeyType="done"
                  testID={TestID.WalletNameInput}
                  textAlign="left"
                  value={unitagInputValue}
                  width={isWeb ? '100%' : undefined}
                  minWidth={unitagNameinputMinWidth}
                  onChangeText={onChangeTextInput}
                  onSubmitEditing={onPressContinue}
                  onLayout={getInitialUnitagNameInputWidth}
                />
                <Text
                  key={UNITAG_SUFFIX}
                  animation="lazy"
                  color="$neutral1"
                  enterStyle={{ opacity: 0, x: SLIDE_IN_AMOUNT }}
                  exitStyle={{ opacity: 0, x: SLIDE_IN_AMOUNT }}
                  fontFamily="$heading"
                  fontSize={isWeb ? fonts.subheading1.fontSize : fontSize}
                  fontWeight="$book"
                  lineHeight={fonts.heading2.lineHeight}
                >
                  {UNITAG_SUFFIX}
                </Text>
              </Flex>
            )}
          </AnimatePresence>
        </AnimatedFlex>
        {unitagAddress && (
          <AnimatedFlex
            row
            alignItems="center"
            gap="$spacing8"
            style={addressViewAnimatedStyle}
            onPress={onPressAddressTooltip}
          >
            <Text color="$neutral2" variant="subheading2">
              {shortenAddress(unitagAddress ?? ADDRESS_ZERO)}
            </Text>
            <TouchableArea onPress={onPressAddressTooltip}>
              <InfoCircleFilled color="$neutral3" size="$icon.20" />
            </TouchableArea>
          </AnimatedFlex>
        )}

        <AvailabilityStatus
          unitagAvailableError={unitagAvailableError}
          addressError={addressError}
          isUnitagAvailable={isUnitagAvailable}
          showTextInputView={showTextInputView}
          mt="$spacing4"
          mb={unitagAddress ? undefined : '$spacing20'}
        />
      </Flex>
      {/* Wrap button in a TouchableArea to add onPress capabilities when the button is disabled. */}
      <TouchableArea disabledStyle={{ cursor: 'default' }} disabled={shouldBlockContinue} onPress={onPressContinue}>
        <Flex row justifyContent="flex-end">
          <Button
            size="large"
            variant="branded"
            isDisabled={shouldBlockContinue || !isUnitagAvailable}
            testID={TestID.Continue}
            loading={showVerificationLoading && isCheckingUnitag} // the validation happens really quickly so only show a loading spinner when the user explicitly tries to continue and we're still checking availability
            onPress={onPressContinue}
          >
            {t('common.button.continue')}
          </Button>
        </Flex>
      </TouchableArea>
      <UnitagInfoModal isOpen={showInfoModal} unitagAddress={unitagAddress} onClose={handleHideInfoModal} />
    </>
  )
}

const animationProps: FlexProps = {
  animation: 'quick',
  enterStyle: { opacity: 0, y: 10 },
}

function AvailabilityStatus({
  unitagAvailableError,
  addressError,
  isUnitagAvailable,
  showTextInputView,
  ...rest
}: {
  unitagAvailableError: string | undefined
  addressError: string | undefined
  isUnitagAvailable: boolean
  showTextInputView: boolean
} & FlexProps): JSX.Element {
  const { t } = useTranslation()
  return (
    <Flex row gap="$spacing8" minHeight={fonts.body2.lineHeight} {...rest}>
      <AnimatePresence>
        {unitagAvailableError || addressError ? (
          <Flex key="error" {...animationProps}>
            <Text key="error" color="$statusCritical" textAlign="center" variant="body2">
              {unitagAvailableError || addressError}
            </Text>
          </Flex>
        ) : isUnitagAvailable && showTextInputView ? (
          <Flex key="available" row alignItems="center" gap="$spacing4" {...animationProps}>
            <CheckmarkCircle color="$accent1" size="$icon.16" />
            <Text textAlign="center" variant="body2">
              {t('unitags.claim.available')}
            </Text>
          </Flex>
        ) : null}
      </AnimatePresence>
    </Flex>
  )
}
