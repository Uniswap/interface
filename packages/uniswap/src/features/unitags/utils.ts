import { TFunction } from 'i18next'
import { UNITAG_VALID_REGEX } from 'uniswap/src/features/unitags/constants'
import { UnitagErrorCodes } from 'uniswap/src/features/unitags/types'

export function parseUnitagErrorCode(t: TFunction, unitag: string, errorCode: UnitagErrorCodes): string {
  switch (errorCode) {
    case UnitagErrorCodes.UnitagNotAvailable:
      return t('unitags.claim.error.unavailable')
    case UnitagErrorCodes.IPLimitReached:
    case UnitagErrorCodes.AddressLimitReached:
    case UnitagErrorCodes.DeviceLimitReached:
      return t('unitags.claim.error.general')
    case UnitagErrorCodes.DeviceActiveLimitReached:
      return t('unitags.claim.error.deviceLimit')
    case UnitagErrorCodes.AddressActiveLimitReached:
      return t('unitags.claim.error.addressLimit')
    default:
      return t('unitags.claim.error.unknown')
  }
}

// Util to handle translations of `yourname`
// If translated string only contains valid Unitag characters, return it lowercased and without spaces
// Otherwise, return 'yourname'
export const getYourNameString = (yourname: string): string => {
  const noSpacesLowercase = yourname.replaceAll(' ', '').toLowerCase()
  if (UNITAG_VALID_REGEX.test(noSpacesLowercase)) {
    return noSpacesLowercase
  }
  return 'yourname'
}
