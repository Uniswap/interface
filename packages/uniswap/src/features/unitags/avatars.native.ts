import { Platform } from 'react-native'
import { getUnitagAvatarUploadUrl, updateUnitagMetadata } from 'uniswap/src/data/apiClients/unitagsApi/UnitagsApiClient'
import { SignMessageFunc } from 'uniswap/src/data/utils'
import { UnitagAvatarUploadCredentials, UnitagGetAvatarUploadUrlResponse } from 'uniswap/src/features/unitags/types'
import { logger } from 'utilities/src/logger/logger'

export function isLocalFileUri(imageUri: string): boolean {
  const localFilePatterns = [
    'file://', // iOS local file prefix
    'content://', // Android Content Provider
    '/storage/', // Android internal storage (absolute path)
    '/data/', // Android internal data storage (absolute path)
  ]

  // Check if the imageUri starts with any of the local file patterns
  return localFilePatterns.some((pattern) => imageUri.startsWith(pattern))
}

export async function uploadFileToS3(
  imageUri: string,
  creds: UnitagAvatarUploadCredentials,
): Promise<{ success: boolean }> {
  if (!creds.preSignedUrl || !creds.s3UploadFields) {
    return { success: false }
  }

  // Standardize the uri for iOS and Android
  const uri = Platform.OS === 'android' ? imageUri : imageUri.replace('file://', '')
  const formData = new FormData()

  // Add the S3 fields to the form data
  Object.entries(creds.s3UploadFields).forEach(([key, value]) => {
    formData.append(key, value)
  })

  // Get the file as a blob to set the Content-Type
  const response = await fetch(uri)
  const blob = await response.blob()
  formData.append('Content-Type', blob.type)

  // Add the file to the form data. We ignore the function signature and input an object with keys uri, type, and name
  // This is the argument that react-native's FormData expects, but for some reason our project thinks it's using typescript's FormData
  // Ignoring the typecheck and forcing the object to be Blob works though
  formData.append('file', {
    uri,
    type: blob.type,
    name: uri,
  } as unknown as Blob)

  // Send the post request to S3 using pre-signed URL and s3 fields
  try {
    const postResponse = await fetch(creds.preSignedUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'multipart/form-data', // Important for S3 to process the file correctly
      },
      body: formData,
    })

    if (!postResponse.ok) {
      throw new Error(`HTTP error! status: ${postResponse.status}`)
    }

    logger.debug('unitags/utils.ts', 'uploadFileToS3', 'Avatar uploaded to S3 successfully')
    return { success: true }
  } catch (error) {
    logger.error(error, {
      tags: { file: 'unitags/utils.ts', function: 'uploadFileToS3' },
    })
    return { success: false }
  }
}

/**
 * Uploads an image to S3 and updates the avatar for a given username and address.
 * Expects imageUri to be a local file, it uploads the file to S3 and updates the avatar URL in the metadata.
 *
 * @param {string} username - The newly claimed unitag.
 * @param {Address} address - The address of the unitag.
 * @param {string} imageUri - The URI of the new avatar image (either a local file or external url).
 * @param {(message: string) => Promise<string>} signMessage - The function to call to sign the message to verify the address.
 * @returns {Promise<boolean>} - A promise that resolves to a boolean indicating whether the avatar was successfully updated.
 */
export async function uploadAndUpdateAvatarAfterClaim({
  username,
  imageUri,
  address,
  signMessage,
}: {
  username: string
  imageUri: string
  address: string
  signMessage: SignMessageFunc
}): Promise<{ success: boolean }> {
  try {
    // First get pre-signedUrl and s3UploadFields from the backend
    const avatarUploadUrlResponse = await getUnitagAvatarUploadUrl({
      data: { username },
      address,
      signMessage,
    })

    // Then upload to S3
    const { success: uploadSuccess } = await uploadFileToS3(imageUri, {
      preSignedUrl: avatarUploadUrlResponse.preSignedUrl,
      s3UploadFields: avatarUploadUrlResponse.s3UploadFields,
    })

    // Check if upload succeeded
    if (!uploadSuccess) {
      return { success: false }
    }

    // Then update profile metadata with the image url
    await updateUnitagMetadata({
      username,
      data: {
        metadata: {
          avatar: avatarUploadUrlResponse.avatarUrl,
        },
        clearAvatar: false,
      },
      address,
      signMessage,
    })
    return { success: true }
  } catch (e) {
    logger.error(e, {
      tags: { file: 'unitags/utils.ts', function: 'uploadAndUpdateAvatarAfterClaim' },
    })
    return { success: false }
  }
}

export const tryUploadAvatar = async ({
  avatarImageUri,
  avatarUploadUrlResponse,
  avatarUploadUrlLoading,
}: {
  avatarImageUri: string | undefined
  avatarUploadUrlResponse: UnitagGetAvatarUploadUrlResponse | undefined
  avatarUploadUrlLoading: boolean
}): Promise<{ success: boolean; skipped: boolean }> => {
  const needsAvatarUpload = !!avatarImageUri && isLocalFileUri(avatarImageUri)
  const isPreSignedUrlReady =
    !avatarUploadUrlLoading && !!avatarUploadUrlResponse?.preSignedUrl && !!avatarUploadUrlResponse?.s3UploadFields
  const shouldTryAvatarUpload = needsAvatarUpload && isPreSignedUrlReady

  if (!shouldTryAvatarUpload) {
    // Return success=true if no upload needed, false if upload needed but can't make request
    return { success: !needsAvatarUpload, skipped: true }
  }

  const avatarUploadResult = await uploadFileToS3(avatarImageUri, {
    preSignedUrl: avatarUploadUrlResponse.preSignedUrl,
    s3UploadFields: avatarUploadUrlResponse.s3UploadFields,
  })

  return { ...avatarUploadResult, skipped: false }
}
