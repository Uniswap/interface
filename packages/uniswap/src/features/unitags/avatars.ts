import { SignMessageFunc } from 'uniswap/src/data/utils'
import { UnitagAvatarUploadCredentials, UnitagGetAvatarUploadUrlResponse } from 'uniswap/src/features/unitags/types'
import { PlatformSplitStubError } from 'utilities/src/errors'

export function isLocalFileUri(_imageUri: string): boolean {
  throw new PlatformSplitStubError('isLocalFileUri')
}

export async function uploadFileToS3(
  _imageUri: string,
  _creds: UnitagAvatarUploadCredentials,
): Promise<{ success: boolean }> {
  throw new PlatformSplitStubError('uploadFileToS3')
}

export async function uploadAndUpdateAvatarAfterClaim(_params: {
  username: string
  imageUri: string
  address: string
  signMessage: SignMessageFunc
}): Promise<{ success: boolean }> {
  throw new PlatformSplitStubError('uploadAndUpdateAvatarAfterClaim')
}

export async function tryUploadAvatar(_params: {
  avatarImageUri: string | undefined
  avatarUploadUrlResponse: UnitagGetAvatarUploadUrlResponse | undefined
  avatarUploadUrlLoading: boolean
}): Promise<{ success: boolean; skipped: boolean }> {
  throw new PlatformSplitStubError('tryUploadAvatar')
}
