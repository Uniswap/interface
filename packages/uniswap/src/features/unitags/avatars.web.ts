import { SignMessageFunc } from 'uniswap/src/data/utils'
import { UnitagAvatarUploadCredentials, UnitagGetAvatarUploadUrlResponse } from 'uniswap/src/features/unitags/types'
import { NotImplementedError } from 'utilities/src/errors'

export function isLocalFileUri(_imageUri: string): boolean {
  throw new Error('Not implemented')
}

export async function uploadFileToS3(
  _imageUri: string,
  _creds: UnitagAvatarUploadCredentials,
): Promise<{ success: boolean }> {
  throw new NotImplementedError('uploadFileToS3')
}

export async function uploadAndUpdateAvatarAfterClaim(_params: {
  username: string
  imageUri: string
  address: string
  signMessage: SignMessageFunc
}): Promise<{ success: boolean }> {
  throw new NotImplementedError('uploadAndUpdateAvatarAfterClaim')
}

export async function tryUploadAvatar(_params: {
  avatarImageUri: string | undefined
  avatarUploadUrlResponse: UnitagGetAvatarUploadUrlResponse | undefined
  avatarUploadUrlLoading: boolean
}): Promise<{ success: boolean; skipped: boolean }> {
  throw new NotImplementedError('tryUploadAvatar')
}
