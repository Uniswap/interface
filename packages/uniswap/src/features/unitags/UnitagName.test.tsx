import { UnitagName } from 'uniswap/src/features/unitags/UnitagName'
import { renderWithProviders } from 'uniswap/src/test/render'

it('renders UnitagName without a name', () => {
  const tree = renderWithProviders(<UnitagName fontSize={12} />)
  expect(tree).toMatchSnapshot()
})

it('renders UnitagName with a name', () => {
  const tree = renderWithProviders(<UnitagName animateIcon fontSize={12} name="testing" />)
  expect(tree).toMatchSnapshot()
})
