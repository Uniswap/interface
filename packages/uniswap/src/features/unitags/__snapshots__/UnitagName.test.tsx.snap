// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UnitagName with a name 1`] = `
<View
  collapsable={false}
  forwardedRef={[Function]}
  jestAnimatedStyle={
    {
      "value": {
        "opacity": 0,
      },
    }
  }
  style={
    {
      "alignItems": "center",
      "alignSelf": "center",
      "flexDirection": "row",
      "gap": 4,
      "opacity": 0,
    }
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "Basel Grotesk",
        "fontSize": 12,
        "fontWeight": "400",
        "lineHeight": 40,
      }
    }
    suppressHighlighting={true}
  >
    testing
  </Text>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    jestAnimatedStyle={
      {
        "value": {
          "opacity": 0,
          "right": -24,
          "top": -4,
          "transform": [
            {
              "translateX": 20,
            },
            {
              "scale": 0.8,
            },
          ],
        },
      }
    }
    style={
      {
        "flexDirection": "row",
        "opacity": 0,
        "position": "absolute",
        "right": -24,
        "top": -4,
        "transform": [
          {
            "translateX": 20,
          },
          {
            "scale": 0.8,
          },
        ],
      }
    }
  />
</View>
`;

exports[`renders UnitagName without a name 1`] = `
<View
  collapsable={false}
  forwardedRef={[Function]}
  jestAnimatedStyle={
    {
      "value": {
        "opacity": 0,
      },
    }
  }
  style={
    {
      "alignItems": "center",
      "alignSelf": "center",
      "flexDirection": "row",
      "gap": 4,
      "opacity": 0,
    }
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "Basel Grotesk",
        "fontSize": 12,
        "fontWeight": "400",
        "lineHeight": 40,
      }
    }
    suppressHighlighting={true}
  />
  <View
    collapsable={false}
    forwardedRef={[Function]}
    jestAnimatedStyle={
      {
        "value": {
          "right": -24,
          "top": -4,
        },
      }
    }
    style={
      {
        "flexDirection": "row",
        "position": "absolute",
        "right": -24,
        "top": -4,
      }
    }
  />
</View>
`;
